vg/VG_Polygon.3
.\" Copyright (c) 2008-2022 Julien Nadeau Carriere <vedge@csoft.net>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd December 21, 2022
.Dt VG_POLYGON 3
.Os Agar 1.7
.Sh NAME
.Nm VG_Polygon
.Nd agar-vg polygon entity
.Sh SYNOPSIS
.Bd -literal
#include <agar/core.h>
#include <agar/gui.h>
#include <agar/vg.h>
.Ed
.Sh DESCRIPTION
.\" IMAGE(/widgets/VG_Polygon.png, "A VG_Polygon")
.Nm
is a
.Xr VG 3
node class which describes a filled polygon with
.Xr VG_Point 3
entities as its vertices.
.Sh INTERFACE
.nr nS 1
.Ft "VG_Polygon *"
.Fn VG_PolygonNew "VG_Node *parent"
.Pp
.Ft "Uint"
.Fn VG_PolygonVertex "VG_Polygon *P" "VG_Point *pt"
.Pp
.Ft "void"
.Fn VG_PolygonDelVertex "VG_Polygon *P" "Uint vtx"
.Pp
.nr nS 0
The
.Fn VG_PolygonNew
function creates a new polygon entity attached to
.Fa parent ,
without any vertices.
.Pp
.Fn VG_PolygonVertex
creates a new vertex using the existing
.Xr VG_Point 3
.Fa pt .
The index of the new vertex is returned.
.Pp
.Fn VG_PolygonDelVertex
removes the specified vertex
.Fa vtx
from the polygon.
The function is a no-op if
.Fa vtx
is an invalid index.
.Pp
.Fn VG_PolygonSetOutline
disables filling of the polygon such that only its outline is displayed.
Default behavior is to display a filled polygon.
.Sh SEE ALSO
.Xr VG 3 ,
.Xr VG_Point 3
.Sh HISTORY
The
.Nm
class first appeared in Agar 1.3.3.


CHANGELOG.md
# Changelog
All notable changes to Agar will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased] - YYYY-MM-DD
### Added
- [**sdl2fb**](https://libagar.org/man3/AG_DriverSDL2FB): New driver for SDL 2.0 (single-window; frame-buffer mode).
- [**sdl2gl**](https://libagar.org/man3/AG_DriverSDL2GL): New driver for SDL 2.0 (single-window; OpenGL mode).
- [**sdl2mw**](https://libagar.org/man3/AG_DriverSDL2MW): New driver for SDL 2.0 (multi-window; OpenGL mode).
- [**AG_Driver**](https://libagar.org/man3/AG_Driver): New operation: `setMouseAutoCapture` (enable, disable or reset mouse auto-capture on button-down).
- [**AG_DriverSw**](https://libagar.org/man3/AG_DriverSw): New option `clampOnResize` (clamp active Agar windows against new display size).
- [**AG_Error**](https://libagar.org/man3/AG_Error): New function `AG_Debug2()` (output debug message when debug level is >= 2).
- [**AG_FontBf**](https://libagar.org/man3/AG_FontBf): New bitmap font engine for the `.agbf` font format.
- [**AG_Menu**](https://libagar.org/man3/AG_Menu): Implemented keyboard shortcuts. Added `AG_MENU_NO_BOOL_MSG` and `AG_MENU_FAST_BOOL_MSG` options.
- [**AG_Menu**](https://libagar.org/man3/AG_Menu): Improved the performance of the `AG_MenuView` renderer by caching state-dependent opaque labels (embedding any icons), so blending is no longer required. Improved appearance of `AG_MenuView`.
- [**AG_Menu**](https://libagar.org/man3/AG_Menu): When an `AG_MenuView` is detached, immediately invalidate the `view` pointers of its `AG_MenuItem`.
- [**AG_Radio**](https://libagar.org/man3/AG_Radio): New `HOMOGENOUS` option to divide space equally between items. Thanks Stephen!
- [**AG_Surface**](https://libagar.org/man3/AG_Surface): Embed 4 general-purpose 16-bit guides. Guides are useful for typography, graphics and other applications.
- [**AG_Text**](https://libagar.org/man3/AG_Text): The surface returned by `AG_TextRender()` now includes Ascent information under `Guide 0`.
- [**AG_Text**](https://libagar.org/man3/AG_Text): New function `AG_TextRenderRTL()` for rendering text right-to-left.
- [**AG_Text**](https://libagar.org/man3/AG_Text): New functions `AG_FontGetStyleName()` and `AG_FontGetStyleByName()`. Convert between the `Uint` and string representations of font styles (including styles, weights and width variants).
- [**AG_Text**](https://libagar.org/man3/AG_Text): New function `AG_FontGetFamilyStyles()`. Return an array of `AG_Font` style flags describing every style, weight and width variant available in a given font's family.
- [**AG_Textbox**](https://libagar.org/man3/AG_Textbox) and [**AG_Editable**](https://libagar.org/man3/AG_Editable): Implement Undo/Redo functionality.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): Draw lines to express the relationship between tree items.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New function `AG_TlistScrollToSelection()`.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New option `AG_TLIST_EXPAND_NODES`. Display nodes (items with children) initially in expanded state.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New option `AG_TLIST_NO_SCALE_ICON`. Disable icon scaling when the icon width exceeds the standard width.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New option `AG_TLIST_FIXED_HEIGHT`. Don't reset the item height on "font-changed".
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New item flag `AG_TLIST_ITEM_DISABLED`. Disable the item individually (prevent selection and display in "disabled" state).
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): Improve keyboard navigation. Allow RIGHT/LEFT keys to Expand/Collapse node items.
- [**AG_Event**](https://libagar.org/man3/AG_Event): New macro [AG_EVENT_DUMP](https://libagar.org/man3/AG_EVENT_DUMP) to produce a listing of `argv[]` arguments on the console.
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): New flag `AG_WIDGET_DISABLE_ON_ATTACH` and functions `AG_PushDisabledState()` and `AG_PopDisabledState`. Make attached child widgets start in DISABLED mode (without raising "widget-disabled").
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): New event _"padding-changed"_. Generated by the style compiler to signal that the padding attribute of a widget has changed.
- [**MAP**](https://libagar.org/man3/MAP): Introduce dynamic Map Objects with [**MAP_Object**](https://libagar.org/man3/MAP_Object). Map Objects coexist with static Map elements and are embedded with the serialized `MAP`. Nodes refer to `MAP_Object` instances using an intermediate `MAP_Location` structure (which adds per-node flags, z-coordinate, height, layer and neighbor information). More than one `MAP` node may refer to a given `MAP_Object`.
- [**MAP**](https://libagar.org/man3/MAP): New Insert Object (`InsertObj`) tool.
- [**MAP**](https://libagar.org/man3/MAP): New functions `MAP_DuplicateLocation()`, `MAP_NodeDelLocation()`, `MAP_NodeDelLocationAtIndex()`.
- [**MAP**](https://libagar.org/man3/MAP): Added height (`h`) and z-coordinate (`z`) attributes to `MAP_Item`. Introduce `MAP_Item` classes.
- [**MAP**](https://libagar.org/man3/MAP): Added Undo/Redo and History Buffer to Map Editor. New function `MAP_ClearHistory()`.
- [**MAP**](https://libagar.org/man3/MAP): Added validity tests to `MAP_Node` and `MAP_Item`.
- [**MAP**](https://libagar.org/man3/MAP): Added persistent Library to the Map Editor. The `pLibs` pointer of `MAP` may be used to specify an alternate `AG_Object` as VFS Root for the Library. Similarly, `pMaps` may be used to specify an alternate VFS Root for loaded Maps.
- `install-agartest.exe` installer for agartest on Windows.
- Provide copies of the OFL as separate files (OFL11.txt and LICENSE.ofl).
- Install a copy of the generated Makefile.config as ${DATADIR}/agar.mk.

vg/vg_polygon.c
/*
 * Copyright (c) 2005-2018 Julien Nadeau Carriere <vedge@csoft.net>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * Polygon entity.
 */

#include <agar/core/core.h>
#include <agar/gui/widget.h>
#include <agar/gui/primitive.h>
#include <agar/gui/checkbox.h>
#include <agar/gui/separator.h>
#include <agar/gui/iconmgr.h>
#include <agar/gui/opengl.h>
#include <agar/vg/vg.h>
#include <agar/vg/vg_view.h>
#include <agar/vg/icons.h>

VG_Polygon *
VG_PolygonNew(void *pNode)
{
	VG_Polygon *ply = Malloc(sizeof(VG_Polygon));

	VG_NodeInit(ply, &vgPolygonOps);
	VG_NodeAttach(pNode, ply);
	return (ply);
}

void
VG_PolygonSetOutline(VG_Polygon *ply, int flag)
{
	ply->outline = flag;
}

Uint
VG_PolygonVertex(VG_Polygon *ply, VG_Point *pt)
{
	VG *vg = VGNODE(ply)->vg;

	AG_ObjectLock(vg);

	ply->pts = Realloc(ply->pts, (ply->nPts + 1)*sizeof(VG_Point *));
	ply->pts[ply->nPts] = pt;
	VG_AddRef(ply, pt);

	AG_ObjectUnlock(vg);
	return (ply->nPts++);
}

void
VG_PolygonDelVertex(VG_Polygon *ply, Uint vtx)
{
	VG *vg = VGNODE(ply)->vg;

	AG_ObjectLock(vg);
	if (vtx < ply->nPts) {
		VG_DelRef(ply, ply->pts[vtx]);
		if (vtx < ply->nPts-1) {
			memmove(&ply->pts[vtx], &ply->pts[vtx+1],
			    (ply->nPts - vtx - 1)*sizeof(VG_Point *));
		}
		ply->nPts--;
	}
	AG_ObjectUnlock(vg);
}

static void
Init(void *_Nonnull obj)
{
	VG_Polygon *vp = obj;

	vp->outline = 0;
	vp->pts = NULL;
	vp->nPts = 0;
	vp->ints = NULL;
	vp->nInts = 0;
}

static int
Load(void *_Nonnull obj, AG_DataSource *_Nonnull ds, const AG_Version *_Nonnull ver)
{
	VG_Polygon *vp = obj;
	Uint i;

	vp->outline = (int)AG_ReadUint8(ds);
	vp->nPts = (Uint)AG_ReadUint8(ds);
	vp->pts = Malloc(vp->nPts*sizeof(VG_Point *));
	for (i = 0; i < vp->nPts; i++) {
		if ((vp->pts[i] = VG_ReadRef(ds, vp, "Point")) == NULL)
			return (-1);
	}
	return (0);
}

static void
Save(void *_Nonnull obj, AG_DataSource *_Nonnull ds)
{
	VG_Polygon *vp = obj;
	Uint i;

	AG_WriteUint8(ds, (Uint8)vp->outline);
	AG_WriteUint8(ds, (Uint8)vp->nPts);
	for (i = 0; i < vp->nPts; i++)
		VG_WriteRef(ds, vp->pts[i]);
}

static void
Destroy(void *_Nonnull obj)
{
	VG_Polygon *vp = obj;

	Free(vp->pts);
	Free(vp->ints);
}

static int
CompareInts(const void *_Nonnull p1, const void *_Nonnull p2)
{
	return (*(const int *)p1 - *(const int *)p2);
}

static void
DrawOutline(VG_Polygon *_Nonnull vp, VG_View *_Nonnull vv)
{
	AG_Color c = VG_MapColorRGB(VGNODE(vp)->color);
	int Ax, Ay, Bx, By, Cx, Cy;
	int i;

	if (vp->nPts < 2) {
		return;
	}
	VG_GetViewCoords(vv, VG_Pos(vp->pts[0]), &Ax,&Ay);
	Cx = Ax;
	Cy = Ay;
	for (i = 1; i < vp->nPts; i++) {
		VG_GetViewCoords(vv, VG_Pos(vp->pts[i]), &Bx,&By);
		AG_DrawLine(vv, Ax,Ay, Bx,By, &c);
		Ax = Bx;
		Ay = By;
	}
	AG_DrawLine(vv, Cx,Cy, Ax,Ay, &c);
}

static void
DrawFB(VG_Polygon *_Nonnull vp, VG_View *_Nonnull vv)
{
	AG_Color c = VG_MapColorRGB(VGNODE(vp)->color);
	int y, x1, y1, x2, y2;
	int ign, miny, maxy;
	int i, i1, i2;
	Uint nPts = vp->nPts;
	int nInts;

	if (vp->ints == NULL) {
		vp->ints = Malloc(nPts*sizeof(int));
		vp->nInts = nPts;
	} else {
		if (nPts > vp->nInts) {
			vp->ints = Realloc(vp->ints, nPts*sizeof(int));
			vp->nInts = nPts;
		}
	}

	/* Find Y maxima */
	VG_GetViewCoords(vv, VG_Pos(vp->pts[0]), &ign, &miny);
	maxy = miny;
	for (i = 1; i < nPts; i++) {
		int vy;
	
		VG_GetViewCoords(vv, VG_Pos(vp->pts[i]), &ign, &vy);
		if (vy < miny) {
			miny = vy;
		} else if (vy > maxy) {
			maxy = vy;
		}
	}

	/* Find the intersections. */
	for (y = miny; y <= maxy; y++) {
		nInts = 0;
		for (i = 0; i < nPts; i++) {
			if (i == 0) {
				i1 = nPts - 1;
				i2 = 0;
			} else {
				i1 = i - 1;
				i2 = i;
			}
			VG_GetViewCoords(vv, VG_Pos(vp->pts[i1]), &ign, &y1);
			VG_GetViewCoords(vv, VG_Pos(vp->pts[i2]), &ign, &y2);
			if (y1 < y2) {
				VG_GetViewCoords(vv, VG_Pos(vp->pts[i1]),
				    &x1, &ign);
				VG_GetViewCoords(vv, VG_Pos(vp->pts[i2]),
				    &x2, &ign);
			} else if (y1 > y2) {
				VG_GetViewCoords(vv, VG_Pos(vp->pts[i1]),
				    &x2, &y2);
				VG_GetViewCoords(vv, VG_Pos(vp->pts[i2]),
				    &x1, &y1);
			} else {
				continue;
			}
			if (((y >= y1) && (y < y2)) ||
			    ((y == maxy) && (y > y1) && (y <= y2))) {
				vp->ints[nInts++] =
				    (((y - y1) << 16) / (y2 - y1)) *
				     (x2 - x1) + (x1 << 16);
			} 
		}

		/* TODO use a less general integer sort */
		qsort(vp->ints, nInts, sizeof(int), CompareInts);

		for (i = 0; i < nInts; i += 2) {
			int xa, xb;

			xa = vp->ints[i] + 1;
			xa = (xa >> 16) + ((xa & 0x8000) >> 15);
			xb = vp->ints[i+1] - 1;
			xb = (xb >> 16) + ((xb & 0x8000) >> 15);
			AG_DrawLineH(vv, xa, xb, y, &c);
		}
	}
}

static void
Draw(void *_Nonnull obj, VG_View *_Nonnull vv)
{
	VG_Polygon *vp = obj;

	if (vp->nPts < 3 || vp->outline) {
		DrawOutline(vp, vv);
		return;
	}
#ifdef HAVE_OPENGL
	if (AGDRIVER_CLASS(WIDGET(vv)->drv)->flags & AG_DRIVER_OPENGL) {
		VG_Color *c = &VGNODE(vp)->color;
		int x1 = WIDGET(vv)->rView.x1;
		int y1 = WIDGET(vv)->rView.y1;
		Uint i;

		glBegin(GL_POLYGON);
		glColor3ub(c->r, c->g, c->b);
		for (i = 0; i < vp->nPts; i++) {
			int x,y;
			VG_GetViewCoords(vv, VG_Pos(vp->pts[i]), &x,&y);
			glVertex2i(x1+x, y1+y);
		}
		glEnd();
	} else
#endif /* HAVE_OPENGL */
	{
		DrawFB(vp, vv);
	}
}

static void
Extent(void *_Nonnull obj, VG_View *_Nonnull vv, VG_Vector *_Nonnull a,
    VG_Vector *_Nonnull b)
{
	VG_Polygon *vp = obj;
	VG_Vector v;
	int i;

	if (vp->nPts < 1) {
		a->x = b->x = 0;
		a->y = b->y = 0;
		return;
	}
	v = VG_Pos(vp->pts[0]);
	a->x = b->x = v.x;
	a->y = b->y = v.y;
	for (i = 0; i < vp->nPts; i++) {
		v = VG_Pos(vp->pts[i]);
		if (v.x < a->x) { a->x = v.x; }
		if (v.y < a->y) { a->y = v.y; }
		if (v.x > b->x) { b->x = v.x; }
		if (v.y > b->y) { b->y = v.y; }
	}
}

static float
PointProximity(void *_Nonnull obj, VG_View *_Nonnull vv, VG_Vector *_Nonnull vPt)
{
	VG_Polygon *vp = obj;
	float d, dMin;
	VG_Vector vInt, A, B, C, m;
	int i;

	if (vp->nPts < 1)
		return (AG_FLT_MAX);

	dMin = AG_FLT_MAX;
	m.x = 0.0f;
	m.y = 0.0f;
	C = A = VG_Pos(vp->pts[0]);
	for (i = 1; i < vp->nPts; i++) {
		B = VG_Pos(vp->pts[i]);
		vInt = *vPt;
		d = VG_PointLineDistance(A, B, &vInt);
		if (d < dMin) {
			dMin = d;
			m = vInt;
		}
		A = B;
	}
	vInt = *vPt;
	d = VG_PointLineDistance(C, A, &vInt);
	if (d < dMin) {
		dMin = d;
		m = vInt;
	}
	if (dMin < AG_FLT_MAX) {
		vPt->x = m.x;
		vPt->y = m.y;
	}
	return (dMin);
}

static void
Delete(void *_Nonnull obj)
{
	VG_Polygon *vp = obj;
	Uint i;

	for (i = 0; i < vp->nPts; i++) {
		if (VG_DelRef(vp, vp->pts[i]) == 0)
			VG_Delete(vp->pts[i]);
	}
}

static void *_Nonnull
Edit(void *_Nonnull obj, VG_View *_Nonnull vv)
{
	VG_Polygon *vp = obj;
	AG_Box *box = AG_BoxNewVert(NULL, AG_BOX_EXPAND);

#ifdef AG_ENABLE_STRING
	AG_LabelNewPolled(box, AG_LABEL_HFILL, _("Points: %d"), &vp->nPts);
#endif
	AG_SeparatorNewHoriz(box);
	AG_CheckboxNewInt(box, 0, _("Render outline"), &vp->outline);
	return (box);
}

VG_NodeOps vgPolygonOps = {
	N_("Polygon"),
	&vgIconPolygon,
	sizeof(VG_Polygon),
	Init,
	Destroy,
	Load,
	Save,
	Draw,
	Extent,
	PointProximity,
	NULL,			/* lineProximity */
	Delete,
	NULL,			/* moveNode */
	Edit
};
